subsys/bluetooth/services/ans.c

```diff
@@ -0,0 +1,496 @@
+	 (IS_ENABLED(CONFIG_BT_ANS_NALRT_CAT_SCHEDULE) << BT_ANS_CAT_SCHEDULE) |                   \
+	 (IS_ENABLED(CONFIG_BT_ANS_NALRT_CAT_HIGH_PRI_ALERT) << BT_ANS_CAT_HIGH_PRI_ALERT) |       \
+	 (IS_ENABLED(CONFIG_BT_ANS_NALRT_CAT_INSTANT_MESSAGE) << BT_ANS_CAT_INSTANT_MESSAGE))
+
+/* Build time ANS supported category bit mask */
+#define BT_ANS_UNALRT_CAT_MASK                                                                     \
+	((IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_SIMPLE_ALERT) << BT_ANS_CAT_SIMPLE_ALERT) |          \
+	 (IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_EMAIL) << BT_ANS_CAT_EMAIL) |                        \
+	 (IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_NEWS) << BT_ANS_CAT_NEWS) |                          \
+	 (IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_CALL) << BT_ANS_CAT_CALL) |                          \
+	 (IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_MISSED_CALL) << BT_ANS_CAT_MISSED_CALL) |            \
+	 (IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_SMS_MMS) << BT_ANS_CAT_SMS_MMS) |                    \
+	 (IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_VOICE_MAIL) << BT_ANS_CAT_VOICE_MAIL) |              \
+	 (IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_SCHEDULE) << BT_ANS_CAT_SCHEDULE) |                  \
+	 (IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_HIGH_PRI_ALERT) << BT_ANS_CAT_HIGH_PRI_ALERT) |      \
+	 (IS_ENABLED(CONFIG_BT_ANS_UNALRT_CAT_INSTANT_MESSAGE) << BT_ANS_CAT_INSTANT_MESSAGE))
+
+/* As per spec, ensure at least one New Alert category is supported */
+BUILD_ASSERT(BT_ANS_NALRT_CAT_MASK != 0,
+	     "At least one ANS New Alert category must be enabled in Kconfig");
+
+static uint16_t alert_sup_cat_bit_mask = BT_ANS_NALRT_CAT_MASK;
+static uint16_t unread_sup_cat_bit_mask = BT_ANS_UNALRT_CAT_MASK;
+
+/* Command ID definitions */
+#define BT_ANS_SEND_ALL_CATEGORY 0xff
+enum bt_ans_command_id {
+	BT_ANS_CTRL_ENABLE_NEW_ALERT,
+	BT_ANS_CTRL_ENABLE_UNREAD,
+	BT_ANS_CTRL_DISABLE_NEW_ALERT,
+	BT_ANS_CTRL_DISABLE_UNREAD,
+	BT_ANS_CTRL_NOTIFY_NEW_ALERT_IMMEDIATE,
+	BT_ANS_CTRL_NOTIFY_UNREAD_IMMEDIATE,
+};
+
+/* Struct definition for Alert Notification Control Point */
+struct alert_ctrl_p {
+	uint8_t cmd_id;
+	uint8_t category;
+} __packed;
+
+/* Struct definition for New Alert */
+struct new_alert {
+	uint8_t category_id;
+	uint8_t num_new_alerts;
+	char text_string[BT_ANS_MAX_TEXT_STR_SIZE];
+} __packed;
+
+/* Struct definition for Unread Alert */
+struct unread_alert_status {
+	uint8_t category_id;
+	uint8_t unread_count;
+} __packed;
+
+/* Mutex for modifying database */
+K_MUTEX_DEFINE(new_alert_mutex);
+K_MUTEX_DEFINE(unread_mutex);
+
+/* Saved messages database */
+static struct new_alert new_alerts[BT_ANS_CAT_NUM];
+static struct unread_alert_status unread_alerts[BT_ANS_CAT_NUM];
+
+/* Initialize to 0, it is control point responsibility to enable once connected */
+static uint16_t alert_cat_enabled_map;
+static uint16_t unread_cat_enabled_map;
+
+/* Notification variables configured by CCC */
+static bool new_alert_notify;
+static bool unread_alert_notify;
```

Comment:
This won't work for `CONFIG_BT_MAX_CONN > 1`. 

Imagine Device A subscribing to alert status, i.e. `unread_alert_notify = (value == BT_GATT_CCC_NOTIFY);` sets `unread_alert_notify` to `true`. 
Then Device B does the same, and `unread_alert_notify` is still set to `true`. 
Then Device A removes its subscription, which sets `unread_alert_notify` to `false`, even when Device B is still subscribed.

You can use `bt_gatt_is_subscribed` to check this instead